### ISSUE TITLE
Realm works slow with encrypted database on x86 architecture CPUs.

### ISSUE THREAD

**Issue by Pasenchuk** [realm/realm-java#2414]
#### Goal

Fast database encryption/decryption on x86 CPUs  

#### Expected Results

Multiple encrypted database connections run with same speed both on Intel and ARM powered devices.

#### Actual Results

App runs really slow on Intel-based devices. Even cheapest ARM-powered tablets (about 30$) provide sufficient performance for our app. But Asus PhonePad (100$) with Intel Atom CPU works really slow. Each screen opens for about 2-3 seconds. Of course, we will optimize the app (now it uses 4 different realm schemas and instances are created in each fragment, activity can have up to 4 fragments on screen). But it is obvious that Intel CPUs cause serious performance issue, because all ARM CPUs works fine.

Note: before enabling database encryption there weren't any performance issues on x86 architecture.

#### Steps & Code to Reproduce

        return new RealmConfiguration.Builder(app)
                .name("chat.realm")
                .encryptionKey(keyStoreHelper.getKey(CHAT_ALIAS))
                .schemaVersion(1)
                .setModules(new ChatModule())
                .build();

#### Version of Realm and tooling

    compile 'io.realm:realm-android:0.87.5'

Android Studio version: 2.0 preview 5

Which Android version and device: 

All with x86 architecture (Asus PhonePad, GenyMotion, plain emulator).

**Comment by beeender** [comment]
@Pasenchuk Could you please let us know which scenario is slow? It would be great if you can make a small sample application to show the slowness.  

PS. I thought it might be caused by the Intel's Houdini (Arm->X86), but apparently it is also slow on the Genymotion. Really strange.

**Comment by Pasenchuk** [comment]
Scenario is not very optimized yet, but earlier (on ARM architecture or everywhere without encryption) everything was good.

1. BaseFragment creates 4 connections to 4 encrypted realm files.
2. Activity can hold up to 4 fragments.
3. Fragment transactions use replace and addToBackstack.
4. Some screens work in separate activities.
5. (Most performance impact) Somewhere on main screen onResume starts sync process. App makes about 17 request to server consequentially with Rx and flatMap operator. Each callback is handled in separate thread, so on this step app opens and closes a lot of realm instances. It takes about 30 seconds on ARM CPU and works in background without any lags and freezes. But x86 CPU can't handle this.
6. If app receives big data array from server (about 2500 entries), writing it to encrypted database takes a long time.

Of course I'll rewrite sync logic, but difference in performance between x86 and ARM is significant.

About GenyMotion: it works fast (because of huge CPU power), but when sync is running (with multiple write transactions), GenyMotion runs slow.

Intel's Houdini (Arm->X86) might cause this issue, because it looks like that only encryption algorithms work slowly on x86. Intel CPUs aren't slower than ARM, but in this case it looks like emulation.

Also I have a question: do realm perform serious calculation on opening transaction? Do I need to minimize number of connections in separate process/fragments?

**Comment by beeender** [comment]
Sorry for the late reply

> Also I have a question: do realm perform serious calculation on opening transaction? Do I need to minimize number of connections in separate process/fragments?
The first time (in all threads) to open Realm might take a bit long time since the Realm needs to verify all schemas, then all schemas will be cached to accelerate the opening process. Also Realm does have a reference counter mechanism to ensure to only open one instance for one thread. It means if you call `getInstance()` twice in the same thread, the same Realm instance will be returned.
No need to minimize minimize number of connections overall, but please do remember to close them otherwise leaks might happen.

Will you please do some profiling with AS to see which Realm call takes significant amount of time?
Also, since in your use case, the Realm writes actually happened in the background thread, which should not block UI for the normal case. However, this might be effect by the scheduler of the system/ IO strategy of the system, it would be hard to say if encryption is slow on that device.

Have you tried the same case without encryption?

Also, Realm's write transaction will block each other. Do you have any sync transaction in the UI thread? That might cause slowness if you have lots of transactions in the background thread because of they lock each other.

**Comment by Pasenchuk** [comment]
>  No need to minimize number of connections overall, but please do remember to close them otherwise leaks might happen.

That is strange, but after moving instances to App class and working only with them (I've implemented reference counting for closing realm instances when all Activities are destroyed) app started to work even faster on Intel CPUs. 
Before I'd always closed instances in onDestroy. Now App closes realms when all Activities are destroyed.

> Have you tried the same case without encryption?

Yes, without encryption everything worked good anywhere.

> Will you please do some profiling with AS to see which Realm call takes significant amount of time?

According to logs, it took some time to open 4 different connections to encrypted db in each fragment. Opening went consequently for all fragments and took almost equal amount of time. So it doesn't look like that all connections used shared instance, may be RC work only for non-encrypted databases.

After moving UI-thread connections to App class, the longest operation is data writing transactions. The speed of putting significant amount of entries (form 50 to 2000 entries for each of 17 tables) have increased from 10-15 seconds to 0.5-3 minutes after enabling encryption. But it doesn't block UI. Also in next data updates with small amount of new entries sync runs pretty fast (10-15 seconds).

> Do you have any sync transaction in the UI thread? 

Yes all UI thread transactions are sync. But it doesn't cause any problems (also most of them are read transactions). Most of write transactions run on background thread.

> lots of transactions in the background thread because of they lock each other.

They run consequently. Also after committing transaction, realm results update themself instantly and changes are shown on UI.

**Comment by beeender** [comment]
> Yes all UI thread transactions are sync. But it doesn't cause any problems (also most of them are read transactions). Most of write transactions run on background thread.

What do you mean read transaction?

**Comment by Pasenchuk** [comment]
Sorry, queries, not transactions. In most cases app doesn't perform inserts in UI thread. And it doesn't happens in main screen, just in specific fragments.

**Comment by jcj52436999** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  jcjacksonengineering@gmail.com
    Domain joe-jackson.net has exceeded the max emails per hour (101/100 (101%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
------ The body of the message is 5013 characters long; only the first
------ 5000 or so are included here.
Received: from github-smtp2-ext7.iad.github.net ([192.30.252.198]:60719 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by rs5.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.86_1)
	(envelope-from <noreply@github.com>)
	id 1aflgF-001eGu-D6
	for jcjackson@joe-jackson.net; Tue, 15 Mar 2016 05:54:23 -0400
Date: Mon, 14 Mar 2016 06:56:45 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1457963805;
	bh=LRn275ERMPTz1E0NgJGeLj9GyartzvlpJhY+na3QuUg=;
	h=From:Reply-To:To:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=OlvAd7nN08C6BbWS0POseSkwdNoO7qKC1s+is2o27Pf/FL83cELhpfoY8Yi5lIY6P
	 Ffb219GdCoqNYFoiPXu4mZdUBroZxsOp5WHX5cwa5OScMc+/jBDySM9cEyMdFimtyp
	 r9ZwJ8f9pWU6GetKknzgwfG57W94obdfCGM2nH28=

**Comment by beeender** [comment]
> The speed of putting significant amount of entries (form 50 to 2000 entries for each of 17 tables) have increased from 10-15 seconds to 0.5-3 minutes after enabling encryption.

This is actually expected slowness when **writing** from our observation as well. The encryption algorithm comes from openssl, We are not sure how much we can improve that in a short time.

More about the reference count, I checked the code again, if all of the Realm instances are closed, the schema will be cleared which leads to re-parse the schema when next time the Realm opened. This should explain your long live Realm instance in application improves the speed. I created an issue to track this https://github.com/realm/realm-java/issues/2465 , I think we can make some improvement on that. No reason to ask you maintain a ref-counter for that :)
